core/prover/src/lib.rs

```diff
@@ -1,687 +1,247 @@
+            return Err(BabyProverError::Internal(
+                "created proof did not pass verification".to_owned(),
+            ));
+        }
 
-            info!("Proof generation is complete");
+        let ret = self
+            .api_client
+            .publish(block, p, prover_data.public_data_commitment);
+        if let Err(e) = ret {
+            return Err(BabyProverError::Api(format!(
+                "failed to publish proof: {}",
+                e
+            )));
+        }
 
-            let full_proof = FullBabyProof {
-                proof: p,
-                inputs: [public_data_commitment],
-                public_data: vec![0 as u8; 10],
-            };
+        info!("finished and published proof for block {}", block);
 
-            let encoded = Self::encode_proof(&full_proof).expect("proof encoding failed");
-            let storage = StorageProcessor::establish_connection()
-                .map_err(|e| format!("establish_connection failed: {}", e))?;
-            storage
-                .store_proof(block_number, &encoded)
-                .map_err(|e| format!("store_proof failed: {}", e))?;
-        } else {
-            thread::sleep(Duration::from_secs(config::PROVER_CYCLE_WAIT));
-        }
         Ok(())
     }
-    pub fn start_timer_interval(&self, rt: &Handle) {
-        let job_ref = self.current_job.clone();
-        rt.spawn(
-            timer::Interval::new_interval(Duration::from_secs(config::PROVER_TIMER_TICK))
-                .fold(job_ref, |job_ref, _| {
-                    let job = job_ref.load(Ordering::Relaxed);
-                    if job > 0 {
-                        //debug!("prover is working on block {}", job);
-                        if let Ok(storage) = StorageProcessor::establish_connection() {
-                            let _ = storage.update_prover_job(job as i32);
-                        }
-                    }
-                    Ok(job_ref)
-                })
-                .map(|_| ())
-                .map_err(|_| ()),
-        )
-        .unwrap();
-    }
 
-    pub fn run(&mut self, shutdown_tx: Sender<()>, stop_signal: Arc<AtomicBool>) {
-        info!("prover is running");
-        while !stop_signal.load(Ordering::SeqCst) {
-            if let Err(err) = self.make_proving_attempt() {
-                error!("Error: {}", err);
+    fn keep_sending_work_heartbeats(&self, start_heartbeats_rx: mpsc::Receiver<(i32, bool)>) {
+        let mut job_id = 0;
+        while !self.stop_signal.load(Ordering::SeqCst) {
+            thread::sleep(self.heartbeat_interval);
+            let (j, quit) = match start_heartbeats_rx.try_recv() {
+                Ok(v) => v,
+                Err(mpsc::TryRecvError::Empty) => (job_id, false),
+                Err(e) => {
+                    panic!("error receiving from hearbeat channel: {}", e);
+                }
+            };
+            if quit {
+                return;
+            }
+            job_id = j;
```

Comment:
why not name `j` `job_id` in the first place

Reply:
`job_id` defined before the loop (scope higher). If start_heartbeats_rx has no data on it, loop should use old value for job id, which is stored in `job_id`.